src/ui/vdom.ts

```diff
@@ -0,0 +1,1481 @@
+/*-----------------------------------------------------------------------------
+| Copyright (c) 2014-2016, PhosphorJS Contributors
+|
+| Distributed under the terms of the BSD 3-Clause License.
+|
+| The full license is in the file LICENSE, distributed with this software.
+|----------------------------------------------------------------------------*/
+
+
+/**
+ * The base attributes object for a virtual DOM node.
+ */
+export
+interface IVNodeAttrs {
+  /**
+   * The key id for the node.
+   *
+   * If a node is given a key id, the generated DOM node will not be
+   * recreated during a rendering update if it only moves among its
+   * siblings in the render tree, provided the type of the node does
+   * not change.
```

Comment:
Perhaps explain why somebody might want this:

> This behavior is an optimization for nodes whose children are known to be static. One example of this usage is the nodes generated for tabs in a tab bar: while their relative order amongst their siblings may change, their content (child nodes) will not change, so they can be implemented as keyed nodes.

Or something of that nature.